File: tests/test_mcp_protocol.py
import asyncio
import json
import os
import sys
from pathlib import Path

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


ROOT = Path(__file__).resolve().parents[1]


async def main() -> None:
    params = StdioServerParameters(
        command=os.environ.get("MCP_PYTHON", sys.executable),
        args=[str(ROOT / "server.py")],
        cwd=str(ROOT),
    )
    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            tools = await session.list_tools()
            names = [tool.name for tool in tools.tools]
            expected = {
                "gazette_status",
                "gazette_categories",
                "search_gazette",
                "get_gazette",
                "browse_gazette",
            }
            assert set(names) == expected, names

            status_result = await session.call_tool("gazette_status", {})
            status = json.loads(status_result.content[0].text)
            assert status["ok"] is True, status
            assert status["status"] == 200, status

            search_result = await session.call_tool(
                "search_gazette", {"category": "gavaaidhu", "max_results": 1}
            )
            search = json.loads(search_result.content[0].text)
            assert search["status"] == 200, search
            assert search["results"], search
            assert search["results"][0]["source"].startswith("https://"), search

            detail_result = await session.call_tool("get_gazette", {"url_or_id": "7539"})
            detail = json.loads(detail_result.content[0].text)
            assert detail["status"] == 200, detail
            assert detail["title"], detail
            assert detail["pdf_url"].endswith(".pdf"), detail

            print(json.dumps({"tools": names, "status": status, "first_result": search["results"][0], "detail": detail}, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    asyncio.run(main())

File: tests/test_mcp_protocol.mjs
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";

const transport = new StdioClientTransport({ command: process.execPath, args: ["dist/server.js"] });
const client = new Client({ name: "maldives-gazette-test", version: "0.2.0" });
await client.connect(transport);
const listed = await client.listTools();
const names = listed.tools.map((tool) => tool.name);
const expected = ["gazette_status", "gazette_categories", "search_gazette", "get_gazette", "browse_gazette", "iulaan_categories", "translate_iulaan_query", "search_iulaan", "get_iulaan", "read_iulaan_attachment"];
for (const name of expected) if (!names.includes(name)) throw new Error(`missing tool: ${name}`);
const status = await client.callTool({ name: "gazette_status", arguments: {} });
const translation = await client.callTool({ name: "translate_iulaan_query", arguments: { query: "admin officer" } });
const search = await client.callTool({ name: "search_iulaan", arguments: { query: "admin officer", announcement_type: "vazeefaa", open_only: true, max_results: 5 } });
console.log(JSON.stringify({ tools: names, status, translation, search }, null, 2));
await client.close();
